### PR TITLE
[IR] Fix continue statement in struct for and add a related test

### LINKED ISSUE #3272: [IR] Continue statement behavior is confusing in struct for
**Describe the bug**
(update) Same issue as #3271. 

Detected by @g1n0st  .
Continue statement in struct for scope will skip the whole lowest activated dense block.
I am not sure whether it is a feature or a bug.

**To Reproduce**
```py
import taichi as ti

ti.init(arch=ti.cuda, kernel_profiler = False)

n = 128
x = ti.field(ti.f32)
block = ti.root.pointer(ti.i, n // 16)
block.dense(ti.i, 16).place(x)

@ti.kernel 
def test():
    x[0] = 1
    x[1] = 2
    cnt = 0
    for i in x:
        if x[i] == 1: continue
        if x[i] == 2: cnt += 1
    print(cnt) # output 0

test()
```

**Log/Screenshots**

```
{11:11}~ ➭ p test.py  
[Taichi] version 0.8.3, llvm 10.0.0, commit 04cf87de, linux, python 3.8.5
[Taichi] Starting on arch=x64
0
```

**Additional comments**
https://github.com/taichi-dev/taichi/blob/3a9c1f3635e3b424bf57e2e5d1de9fd909097820/taichi/ir/statements.cpp#L7
https://github.com/taichi-dev/taichi/blob/master/taichi/codegen/codegen_llvm.cpp#L802
A continue statement will emit a return statement in LLVM.

https://github.com/taichi-dev/taichi/blob/master/taichi/codegen/codegen_llvm.cpp#L1527
But every thread in struct for will loop over the whole lowest dense block.
I guess that a return statement in the kernel will force the thread to jump over the owest dense block loop, and resulting in this confusing skipping behavior.

### PR DESCRIPTION
**Pull request by BillXu2000**
fixes #3272

It seems that only a few backends (including llvm, metal) support struct-for, and I only fix this problem for llvm CUDA and cpu backends.
Not sure whether this works for metal.

### CODE REVIEW

**Review by g1n0st — APPROVED** [review]
LGTM

File: taichi/codegen/codegen_llvm.cpp [hunk 1]
```diff
@@ -798,7 +798,11 @@ void CodeGenLLVM::visit(WhileControlStmt *stmt) {
 
 void CodeGenLLVM::visit(ContinueStmt *stmt) {
   using namespace llvm;
-  if (stmt->as_return()) {
+  auto stmt_in_range_for = [&]() {
+    auto *offl = stmt->scope->cast<OffloadedStmt>();
+    return offl->task_type == OffloadedStmt::TaskType::range_for;
+  };
+  if (stmt->as_return() && stmt_in_range_for()) {
```
[the remark this answers is not in the document]
Comment (BillXu2000) [reply taichi/codegen/codegen_llvm.cpp]: removed

File: taichi/backends/metal/codegen_metal.cpp [hunk 2]
```diff
@@ -679,7 +679,16 @@ class KernelCodegenImpl : public IRVisitor {
   }
 
   void visit(ContinueStmt *stmt) override {
-    if (stmt->as_return()) {
+    auto stmt_in_off_for = [&]() {
```
Comment (k-ye) [inline taichi/backends/metal/codegen_metal.cpp:682]:
```suggestion
    auto stmt_in_offl_for = [stmt]() {
```

File: taichi/backends/vulkan/codegen_vulkan.cpp [hunk 3]
```diff
@@ -995,7 +995,16 @@ class TaskCodegen : public IRVisitor {
   }
 
   void visit(ContinueStmt *stmt) override {
-    if (stmt->as_return()) {
+    auto stmt_in_off_for = [&]() {
```
Comment (k-ye) [inline taichi/backends/vulkan/codegen_vulkan.cpp:998]:
```suggestion
    auto stmt_in_offl_for = [stmt]() {
```

File: taichi/codegen/codegen_llvm.cpp [hunk 4]
```diff
@@ -798,7 +798,16 @@ void CodeGenLLVM::visit(WhileControlStmt *stmt) {
 
 void CodeGenLLVM::visit(ContinueStmt *stmt) {
   using namespace llvm;
-  if (stmt->as_return()) {
+  auto stmt_in_off_range_for = [&]() {
```
Comment (k-ye) [inline taichi/codegen/codegen_llvm.cpp:801]:
```suggestion
  auto stmt_in_offl_range_for = [stmt]() {
```

**Review by k-ye — APPROVED** [review]
LGTM!

**Pull request merged**